btcec/schnorr/musig2/keys.go

```diff
@@ -0,0 +1,145 @@
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+)
+
+var (
+	// KeyHashTag is the tagged hash tag used to compute the hash of the
+	// list of sorted public keys.
+	KeyHashTag = []byte("KeyAgg list")
+
+	// KeyAggTag is the tagged hash tag used to compute the key aggregation
+	// coefficient for each key.
+	KeyAggTag = []byte("KeyAgg coefficient")
+)
+
+// sortKeys takes a set of schnorr public keys and returns a new slice that is
+// a copy of the keys sorted in lexicographical order bytes on the x-only
+// pubkey serialization.
+func sortKeys(keys []*btcec.PublicKey) []*btcec.PublicKey {
+	keyCopy := make([]*btcec.PublicKey, len(keys))
+	copy(keyCopy, keys)
+
+	sort.Slice(keyCopy, func(i, j int) bool {
+		keyIBytes := schnorr.SerializePubKey(keyCopy[i])
+		keyJBytes := schnorr.SerializePubKey(keyCopy[j])
+
+		return bytes.Compare(keyIBytes, keyJBytes) == -1
+	})
+
+	return keyCopy
+}
+
+// keyHashFingerprint computes the tagged hash of the series of (sorted) public
+// keys passed as input. This is used to compute the aggregation coefficient
+// for each key. The final computation is:
+//   * H(tag=KeyAgg list, pk1 || pk2..)
+func keyHashFingerprint(keys []*btcec.PublicKey) []byte {
+	var keyBytes bytes.Buffer
+	for _, key := range keys {
+		keyBytes.Write(schnorr.SerializePubKey(key))
+	}
+
+	h := chainhash.TaggedHash(KeyHashTag, keyBytes.Bytes())
+	return h[:]
+}
+
+// isSecondKey returns true if the passed public key is the second key in the
+// (sorted) keySet passed.
+func isSecondKey(keySet []*btcec.PublicKey, targetKey *btcec.PublicKey) bool {
+	// For this comparison, we want to compare the raw serialized version
+	// instead of the full pubkey, as it's possible we're dealing with a
+	// pubkey that _actually_ has an odd y coordinate.
+	equalBytes := func(a, b *btcec.PublicKey) bool {
+		return bytes.Equal(
+			schnorr.SerializePubKey(a),
+			schnorr.SerializePubKey(b),
+		)
+	}
+
+	for i := range keySet {
+		if !equalBytes(keySet[i], keySet[0]) {
+			return equalBytes(keySet[i], targetKey)
+		}
+	}
+
+	return false
+}
+
+// aggregationCoefficient computes the key aggregation coefficient for the
+// specified target key. The coefficient is computed as:
+//  * H(tag=KeyAgg coefficient, keyHashFingerprint(pks, pk) || pk)
+func aggregationCoefficient(keySet []*btcec.PublicKey,
+	targetKey *btcec.PublicKey) *btcec.ModNScalar {
+
+	var mu btcec.ModNScalar
+
+	// If this is the second key, then this coefficient is just one.
+	//
+	// TODO(roasbeef): use intermediate cache to keep track of the second
+	// key, can just store an index, otherwise this is O(n^2)
+	if isSecondKey(keySet, targetKey) {
+		return mu.SetInt(1)
+	}
+
+	// Otherwise, we'll compute the full finger print hash for this given
+	// key and then use that to compute the coefficient tagged hash:
+	//  * H(tag=KeyAgg coefficient, keyHashFingerprint(pks, pk) || pk)
+	var coefficientBytes bytes.Buffer
+	coefficientBytes.Write(keyHashFingerprint(keySet))
```

Comment:
Not sure if you intend this to follow the same API as some other impl (I'm not familiar with the existing musig impls), but this can be trivially moved to `AggregateKeys` to avoid repeating the hash.

For reference, here's my test impl for 3 round musig: https://github.com/matheusd/mrttree/blob/master/primitives.go#L80-L99

Reply:
Yep, as noted in the PR body I've done zero optimizations as is. My plan is to memoize/cache all this information in an internal signing/verification context.

Reply:
The other thing that can make things slower for really a really large number of signers is that the `isSecondKey` function ends up being O(N^2). Instead we can compute that once, then just record the key index and eliminate all the looping/serialization.